README.md
<p align="center">
    <img style="display:block;text-align:center" src="./docs/readme-resources/jquery-mobile-seed.svg" alt="logo-text" width="600" />
    <br/>
    <h1 align="center">jQuery Mobile Seed</h1>
    <p align="center" style="font-size: 1.2rem;">A quick start kit for jQuery Mobile Framework based website development with SASS support</p>
</p>

<!-- Badges -->
<p align="center">
  <a href="https://travis-ci.org/apareciumlabs/jquery-mobile-seed">
    <img src="https://travis-ci.org/apareciumlabs/jquery-mobile-seed.svg?branch=master" alt="Build Status" height="18">
  </a>
  <a href="#contributors">
    <img src="https://img.shields.io/badge/all_contributors-1-orange.svg" alt="All Contributors" height="18">
  </a>
  <a href="https://david-dm.org/apareciumlabs/jquery-mobile-seed">
    <img src="https://david-dm.org/apareciumlabs/jquery-mobile-seed/status.svg" alt="dependencies Status" height="18">
  </a>
  <a href="https://david-dm.org/apareciumlabs/jquery-mobile-seed?type=dev">
    <img src="https://david-dm.org/apareciumlabs/jquery-mobile-seed/dev-status.svg" alt="devDependencies Status" height="18">
   </a>
  <a href="LICENSE.md">
    <img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT" height="18">
  </a>
  <a href="https://codecov.io/gh/apareciumlabs/jquery-mobile-seed">
  <img src="https://codecov.io/gh/apareciumlabs/jquery-mobile-seed/branch/master/graph/badge.svg" />
</a>
</p>

This repository could be used by anyone who's looking for a quick starter kit for jQuery Mobile framework based website development. This boilerplate uses [Gulp](https://gulpjs.com/) for building and automating the workflow. Following are some key features we've included for you.

* The most recent version of [jQuery Mobile 1.4.5](http://demos.jquerymobile.com/1.4.5/)
* Usage of [Sass: Syntactically Awesome Style Sheets](https://sass-lang.com/) over CSS.
* Ready to go build system using Gulp.
* Simplified development process with npm scripts.
* Ability to use your favorite bower packages.
* Organized easy to use folder structure for beginners.

# Quick Links

[Demo](https://apareciumlabs-jquery-mobile-seed.netlify.com/) | [Documentation](http://demos.jquerymobile.com/1.4.5/) | [Theme Roller](https://themeroller.jquerymobile.com/)
|---|---|---|

# Quick start

> The generated project have dependencies that require `node` together with `npm` & `bower`.

**Make sure you have [Node](https://nodejs.org/en/download/) version >= 8.0, [npm](https://www.npmjs.com/) >= 5 or [Yarn](https://yarnpkg.com) and [Bower](https://bower.io/) >= 1.8**

```bash
# clone our repository
# --depth 1 removes all but one .git commit history
git clone --depth 1 https://github.com/apareciumlabs/jquery-mobile-seed.git

# change the directory
cd jquery-mobile-seed

# install the bower dependencies
bower install

# install the dependencies with npm
npm install

# start the development server
npm start

```

Once the dev server is fired up, it'll automatically open up a new tab. If not, navigate to [http://0.0.0.0:3000](http://0.0.0.0:3000) or [http://localhost:3000](http://localhost:3000) manually in your browser.

# Table of Contents

* [File Structure](#file-structure)
* [Getting Started](#getting-started)
  * [Prerequisites](#prerequisites)
  * [Setting Up](#setting-up)
  * [Running the app](#running-the-app)
* [Configuration](#configuration)
  * [Add third-party dependencies](#add-third-party-dependencies)
* [Styling](#styling)
* [Deployment](#deployment)
  * [Netlify](#netlify)
* [Built With](#built-with)
* [Contributing](#contributing)
* [Contributors](#contributors)
* [License](#license)

# File Structure

```
jquery-mobile-seed/
 ├── docs/                          * contains documents and document resources
 ├── node_modules/                  * contains dependencies pulled from npm
 ├── src/                           * styles folder
 │   ├── assets/                    * static assets such as images, icons, fonts goes here
 │   ├── scss/                      * styles folder
 │   │   ├── partials/              * place all the sass partial stylesheets in this folder
 │   │   └── styles.scss            * the main stylesheet for the project which gets compiled to CSS
 │   ├── scripts/                   * custom javascript script files
 │   ├── vendor/                    * third party bower libraries will be copied here
 │   └── index.html                 * entry HTML file.
 ├── .all-contributorsrc            * contains info ablout repo contributors
 ├── .bowerrc                       * bower config file
 ├── .dockerignore                  * contains files that are ignored from docker
 ├── .editorconfig                  * helps define and maintain consistent coding styles between different editors and IDEs
 ├── .eslintrc                      * ecmascript linting configuration file
 ├── .gitignore                     * contains files that are ignored from git
 ├── .netlify.toml                  * netlify config file
 ├── .npmrc                         * npm config file to house project wide custom configs
 ├── .nvmrc                         * node version manager config file
 ├── .pullapprove.yml               * pullapprove config file
 ├── .sass-lint.yml                 * sass linting configuration file
 ├── .travis.yml                    * travis ci configuration file
 ├── bower.json                     * contains bower dependencies
 ├── CONTRIBUTING.md                * project contributing guidelines
 ├── docker-compose.yml             * docker compose file
 ├── Dockerfile                     * docker config file
 ├── gulpfile.js                    * main buld configuration file. contains all the gulp tasks.
 ├── LICENSE.md                     * licensing information
 ├── package.json                   * contains all the npm scripts for building, running, deploying etc. and contains all the dependencies
 └── README.md                      * Readme file for the repository

```

# Getting Started

## Prerequisites

What you need to run this app:

* The generated project have dependencies that require `node` together with `npm` & `bower`.
* Ensure you're running the latest stable versions Node, NPM and Bower.

>Make sure you have `Node` and `NPM` installed by running simple commands on the command line to see what version of each is installed.

* Node - Type `node -v` on the terminal.
* NPM -  Type `npm -v` on the terminal.
* Bower -  Type `bower -v` on the terminal.

If you do not have them installed, click [here](https://nodejs.org/en/download/) and grab the latest stable version of `node` and `npm` will be automatically installed along with it. Or if you have `brew` already installed in your local machine, execute `brew install node` command to get `node`.

If you have npm installed, It's really easy to grab the latest version of Bower by executing `npm install -g bower` which will install it globally.

Though the project is built on top of gulp, we've created npm scripts to run gulp tasks. So you don't have to install gulp or gulp CLI globally on your working machine.

## Setting Up

* `clone` the repository
* `cd jquery-mobile-seed` to change the directory
* `bower install` to install the bower dependencies
* `npm install` to install the dependencies with npm

## Running the app

After you have installed all dependencies you can now run the app. Run `npm start` to start a local server using `gulp` which will watch your stylesheets and javascript files for changes, compile, minify, build and reload the browser automatically using `browser-sync` library.
The dev server will be opened in a new tab and usually on http://localhost:3000 and the Access URLs will be displayed on the terminal.

### server

```bash
# development
npm start

# production
npm run server:prod
```

# Configuration

The `gulp` tasks are defined inside the `gulpfile.js` file found on the root of the application and you need to extend this file if you want to customize the application.

## Add third-party dependencies

To use any third party libraries other than the included `font-awesome` and `jquery` libs, find the package you want in the [bower package repository](https://bower.io/search/) and use the following command to add it to your project.

```bash
bower install $package --save
```

Take a look at the bellow example:

Lets say that you want to add `moment.js` to your project. Just run the command bellow.

```bash
bower install moment --save
```

NOTE: Some packages won't work as expected and you might have to do overrides in the `bower.json` file.

# Styling

The `styles.scss` file inside the `sass` directory is the main stylesheet for the project and will be compiled and minified into an external `.css` and is embedded in the `index.html` file.
If you want to add your own stylesheet, we recommend that you place it under the `scss/partials` folder and import it in the `styles.scss` file.

For example if you want to include the styles for a slider:
1) Create a `_slider.scss` partial file in the `scss/partials` directory.
2) In `styles.scss` add `@import 'partials/slider.scss';`

For the purpose of the demo, we have overriden the native look and feel of some jQuery mobile components. Feel free to remove them by locating in partials.

# Testing

Execute the following command to run your unit tests. We use [codecov](https://codecov.io/) package to generate test coverage reports and the generated reports can be found in the `coverage` folder.

```bash
npm run test
```

# Linting

Execute the following commands to generate linting for styles and scripts.

```bash
# all
npm run lint

# javascript
npm run lint:scripts

# sass
npm run lint:styles
```

# Deployment

## Building the app

### For Development

Execute the following command to build your files in the development mode. A new folder called `.temp` will be created and the artifacts will be saved there.

```bash
npm run build:dev
```

### For Production

Execute the following command to build your files in the production mode. A new folder called `dist` will be created and the artifacts will be saved there.

```bash
npm run build:prod
```

## Deploy to Netlify

Deploy to [Netlify](https://www.netlify.com/) using this template with one single click. It has an awesome free plan and a nifty [CLI](https://github.com/netlify/netlify-cli) which makes the deployment process much easier.

Click [here](https://apareciumlabs-jquery-mobile-seed.netlify.com/) to see the deployed app.

<!-- Markdown snippet -->
[![Deploy to Netlify](https://www.netlify.com/img/deploy/button.svg)](https://app.netlify.com/start/deploy?repository=https://github.com/apareciumlabs/jquery-mobile-seed)

# Built With

<a href="https://jquerymobile.com"><img src="./docs/readme-resources/jquery-mobile.svg" alt="npm" height="20" /></a>&nbsp;&nbsp;
<a href="https://www.npmjs.com/"><img src="./docs/readme-resources/npm.svg" alt="npm" height="20" /></a>&nbsp;&nbsp;
<a href="https://bower.io/"><img src="./docs/readme-resources/bower.svg" alt="bower" height="30" /></a>&nbsp;&nbsp;
<a href="https://sass-lang.com/"><img src="./docs/readme-resources/sass.svg" alt="sass" height="30" /></a>&nbsp;&nbsp;
<a href="https://gulpjs.com/"><img style="display:inline-block;margin: 5px 10px" src="./docs/readme-resources/gulp.svg" alt="gulp" height="30" /></a>

# Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for contributing guidelines and to learn about our code of conduct.

# Contributors

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore -->
| [<img src="https://avatars3.githubusercontent.com/u/25959096?v=4" width="80px;"/><br /><sub><b>Brion Mario</b></sub>](http://www.brionmario.com/)<br />[💻](https://github.com/apareciumlabs/jquery-mobile-seed/commits?author=brionmario "Code") [📖](https://github.com/apareciumlabs/jquery-mobile-seed/commits?author=brionmario "Documentation") [🐛](https://github.com/apareciumlabs/jquery-mobile-seed/issues?q=author%3Abrionmario "Bug reports") [⚠️](https://github.com/apareciumlabs/jquery-mobile-seed/commits?author=brionmario "Tests") |
| :---: |
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/kentcdodds/all-contributors) specification. Contributions of any kind welcome!

# License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details.


CHANGELOG.md
<a name="1.0.1-alpha.0"></a>
## 1.0.1-alpha.0 (2018-12-07)


### Bug Fixes

* **build:** fix bower install error :green_heart: ([a06f59c](https://github.com/apareciumlabs/jquery-mobile-seed/commit/a06f59c))





src/index.php
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="msapplication-TileColor" content="#ffffff">
  <meta name="theme-color" content="#ffffff">
  <meta name="author" content="Brion Silva">
  <title>jQuery Mobile Seed by Aparecium Labs</title>
  <!-- Favicon Package -->
  <link rel="apple-touch-icon" sizes="180x180" href="./assets/icons/favicon_package/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="./assets/icons/favicon_package/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="./assets/icons/favicon_package/favicon-16x16.png">
  <link rel="manifest" href="./assets/icons/favicon_package/site.webmanifest">
  <link rel="mask-icon" href="./assets/icons/favicon_package/safari-pinned-tab.svg" color="#5bbad5">
  <!-- inject:css -->
  <!-- endinject -->
</head>

<body>
<div data-role="page">
  <?php require './sidebar.php'?><!-- /panel -->
  <?php require './header.php'?><!-- header -->
  <div role="main" class="overlay ui-content main-content">
    <div class="ui-grid-a ui-responsive">
      <div class="ui-block-a">
        <section class="left-section">
          <div class="display-table">
            <div class="display-table-cell">
              <div class="content">
                <div class="aparecium-logo-container">
                  <svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" width="1200px" height="1200px" viewBox="0 0 1200 1200" enable-background="new 0 0 1200 1200" xml:space="preserve" class="aparecium-logo fox"><g id="Fox"><polygon fill="#BDBEC0" stroke="#FFFFFF" stroke-miterlimit="10" points="1058.977,356.682 1100.332,463.247 926.018,623.893 926.214,567.694 	"/><polygon fill="#F15F3E" stroke="#FFFFFF" stroke-miterlimit="10" points="837.496,532.569 872.783,588.951 716.814,817.89 	"/><polygon fill="#BDBEC0" stroke="#FFFFFF" stroke-miterlimit="10" points="136.135,354.199 292.34,568.224 270.073,626.543 89.287,460.594 	"/><polygon fill="#1B2635" stroke="#FFFFFF" stroke-miterlimit="10" points="1058.977,122.343 1030.348,332.029 926.214,567.694 1058.977,356.682 	"/><polygon fill="#202E42" stroke="#FFFFFF" stroke-miterlimit="10" points="945.253,338.657 1058.977,122.343 1030.348,332.029 926.214,567.694 	"/><polyline fill="#364867" stroke="#FFFFFF" stroke-miterlimit="10" points="721.387,343.694 1058.977,122.343 945.253,338.657 721.387,343.694 	"/><polyline fill="#2C3B56" stroke="#FFFFFF" stroke-miterlimit="10" points="721.387,343.694 926.214,567.694 945.253,338.657 721.387,343.694 	"/><polygon fill="#2C3B56" stroke="#FFFFFF" stroke-miterlimit="10" points="136.469,121.814 252.579,338.657 292.34,568.224 174.11,350.851 	"/><polygon fill="#1C2635" stroke="#FFFFFF" stroke-miterlimit="10" points="136.469,121.814 174.11,350.851 292.34,568.224 136.135,354.199 	"/><polygon fill="#364867" stroke="#FFFFFF" stroke-miterlimit="10" points="136.469,121.814 472.603,343.699 252.579,338.657 	"/><polygon fill="#2C3B56" stroke="#FFFFFF" stroke-miterlimit="10" points="292.869,567.434 471.556,344.886 472.603,343.699 253.107,337.87 	"/><polygon fill="#F58846" stroke="#FFFFFF" stroke-miterlimit="10" points="370.276,532.437 472.603,343.699 292.34,568.224 	"/><polygon fill="#F05A2B" stroke="#FFFFFF" stroke-miterlimit="10" points="596.996,615.94 721.387,343.694 598.782,411.289 472.603,343.699 	"/><polygon fill="#C74227" stroke="#FFFFFF" stroke-miterlimit="10" points="721.387,343.694 472.603,343.699 598.782,411.289 	"/><polygon fill="#F58846" stroke="#FFFFFF" stroke-miterlimit="10" points="721.387,343.694 926.214,567.694 837.496,532.569 	"/><polygon fill="#F36B2F" stroke="#FFFFFF" stroke-miterlimit="10" points="721.387,343.694 837.496,532.569 716.814,817.89 598.916,751.268 487.179,817.89 370.276,532.437 472.603,343.699 596.996,615.94 	"/><polygon fill="#F89C68" stroke="#FFFFFF" stroke-miterlimit="10" points="292.34,568.224 370.276,532.437 352.78,588.951 	"/><polygon fill="#F15E3E" stroke="#FFFFFF" stroke-miterlimit="10" points="370.276,532.437 487.179,817.89 352.78,588.951 	"/><polyline fill="#D34E29" stroke="#FFFFFF" stroke-miterlimit="10" points="292.34,568.224 352.78,588.951 487.179,817.89 292.34,568.224 	"/><polygon fill="#0B1525" stroke="#FFFFFF" stroke-miterlimit="10" points="292.34,568.224 270.868,626.277 487.179,817.89 	"/><polygon fill="#F6921E" stroke="#FFFFFF" stroke-miterlimit="10" points="487.179,817.89 598.916,751.268 716.814,817.89 	"/><polygon fill="#0B1525" stroke="#FFFFFF" stroke-miterlimit="10" points="487.179,817.89 716.814,817.89 596.928,972.406 	"/><polygon fill="#F15A2B" stroke="#FFFFFF" stroke-miterlimit="10" points="716.814,817.89 823.181,900.646 596.928,972.406 	"/><polygon fill="#F15A2B" stroke="#FFFFFF" stroke-miterlimit="10" points="487.179,817.89 596.928,972.406 365.107,898.654 	"/><path fill="#DD5432" stroke="#FFFFFF" stroke-miterlimit="10" d="M596.928,972.406c1.191,0,225.855-72.16,225.855-72.16 l-16.302,54.875l-105.375,60.041L596.928,972.406z"/><polygon fill="#DD5432" stroke="#FFFFFF" stroke-miterlimit="10" points="365.107,898.654 596.928,972.406 485.988,1015.162 377.831,956.907 	"/><polygon fill="#E6E7E9" stroke="#FFFFFF" stroke-miterlimit="10" points="596.928,972.406 486.385,1015.559 548.018,1069.639 590.168,1043.791 635.099,1067.652 701.106,1014.764 	"/><polygon fill="#DEDEE0" stroke="#FFFFFF" stroke-miterlimit="10" points="89.287,460.594 487.179,817.89 9.228,668.955 	"/><polygon fill="#DEDEE0" stroke="#FFFFFF" stroke-miterlimit="10" points="716.814,817.89 1100.332,463.247 1184.629,662.065 	"/><polygon fill="#E9EAEB" stroke="#FFFFFF" stroke-miterlimit="10" points="716.814,817.89 1183.038,662.86 1027.961,825.094 	"/><polygon fill="#EDEDED" stroke="#FFFFFF" stroke-miterlimit="10" points="716.814,817.89 818.807,900.646 1028.757,826.686 	"/><polygon fill="#EDEDED" stroke="#FFFFFF" stroke-miterlimit="10" points="153.963,825.094 487.179,817.89 365.107,898.654 	"/><path fill="#C74227" stroke="#FFFFFF" stroke-miterlimit="10" d="M482.409,1015.958"/><polygon fill="#E9EAEB" stroke="#FFFFFF" stroke-width="0.25" stroke-miterlimit="10" points="9.228,668.955 153.963,825.094 487.179,817.89 	"/><polygon fill="#D34E29" stroke="#FFFFFF" stroke-miterlimit="10" points="872.783,588.951 926.214,567.694 716.814,817.89 	"/><polygon fill="#0B1525" stroke="#FFFFFF" stroke-miterlimit="10" points="926.214,567.694 926.018,623.893 716.814,817.89 	"/><polygon fill="#F89C68" stroke="#FFFFFF" stroke-miterlimit="10" points="872.783,588.951 837.496,532.569 926.214,567.694 	"/></g></svg>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
      <div class="ui-block-b">
        <section class="right-section">
          <div class="display-table">
            <div class="display-table-cell">
              <div class="content">
                <h2 class="title">
                  <img src="assets/img/jquery-mobile.svg" class="jquery-mobile-logo">
                </h2>
                <p class="desc">We <span style="color: red">&hearts;</span> open-source and always trying our best to
                  contribute back to the community.
                  Please Subscribe to our newsletter to stay up to date with more awesome projects like this.
                </p>
                <div class="email-subscription-container" id="mc_embed_signup">
                  <form
                    action="https://apareciumlabs.us15.list-manage.com/subscribe/post?u=41620257667e0123ce58b4277&amp;id=c1c1df411f"
                    method="post" id="mc-embedded-subscribe-form" name="mc-embedded-subscribe-form" target="_blank"
                    novalidate>
                    <div id="mc_embed_signup_scroll">
                      <input type="email"
                             class="email-input"
                             value=""
                             name="EMAIL"
                             id="mce-EMAIL"
                             placeholder="Enter your email"
                             required/>
                      <!-- real people should not fill this in and expect good things - do not remove this or
                      risk form bot -->
                      <div style="position: absolute; left: -5000px;" aria-hidden="true">
                        <input
                          type="text"
                          name="b_41620257667e0123ce58b4277_c1c1df411f"
                          tabindex="-1"
                          value=""/>
                      </div>
                      <button id="mc-embedded-subscribe" class="subscribe-btn" name="subscribe" type="submit">
                        <b>Subscribe</b>
                      </button>
                    </div>
                  </form>
                </div>
                <p class="post-desc">Please note that this subscription can be cancelled at any time.</p>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div><!-- /grid-a -->
  </div><!-- /content -->
  <?php require './footer.php'?><!--footer -->
</div><!-- page -->
<!-- inject:js -->
<!-- endinject -->
</body>
</html>


gulpfile.js
const gulp = require('gulp');
const babel = require('gulp-babel');
const concat = require('gulp-concat');
const inject = require('gulp-inject');
const gutil = require('gulp-util');
const imagemin = require('gulp-imagemin');
const autoprefixer = require('gulp-autoprefixer');
const del = require('del');
const htmlmin = require('gulp-htmlmin');
const sass = require('gulp-sass');
const header = require('gulp-header');
const cleanCSS = require('gulp-clean-css');
const rename = require("gulp-rename");
const sassLint = require('gulp-sass-lint');
const eslint = require('gulp-eslint');
const pkg = require('./package.json');
const browserSync = require('browser-sync');
const uglify = require('gulp-uglify');
const sourcemaps = require('gulp-sourcemaps');
const wiredep = require('wiredep');
const urlAdjuster = require('gulp-css-url-adjuster');
const bowerlibs = require('main-bower-files');
const phpConnect = require('gulp-connect-php');

/**
 * Browser Support declaration
 * @type {string[]}
 */
const AUTO_PREFIX_BROWSERS = [
  'ie >= 10',
  'ie_mob >= 10',
  'ff >= 30',
  'chrome >= 34',
  'safari >= 7',
  'opera >= 23',
  'ios >= 7',
  'android >= 4.4',
  'bb >= 10'
];

/**
 * Banner for the compiled CSS files
 * @type {string}
 */
const banner = [
  '/*! ========================================================================= *\n',
  ' *  <%= pkg.title %> \n',
  ' *  ========================================================================= *\n',
  ' *  Version: <%= pkg.version %>\n',
  ' *  Author: <%= pkg.author %>\n',
  ' *  ========================================================================= *\n',
  ' *  GitHub Repo: <%= pkg.repository.url %>\n',
  ' *  ========================================================================= *\n',
  ' *  Copyright Ⓒ ' + (new Date()).getFullYear(), '\n',
  ' *  Licensed under <%= pkg.license %>\n',
  ' *  ========================================================================= *\n',
  ' */\n\n',
  ''
].join('');

const CONFIG = {
  filenames: {
    dev: {
      scripts: 'main.js',
      vendorJS: 'vendor.js',
      vendorCSS: 'vendor.css',
      styles: 'styles.css',
    },
    prod: {
      scripts: 'main.bundle.js',
      vendorJS: 'vendor.bundle.js',
      vendorCSS: 'vendor.bundle.css',
      styles: 'styles.bundle.css',
    }
  },
  paths: {
    dev: {
      root: '.temp',
      scripts: '.temp/js',
      styles: '.temp/css',
      pages: '.temp/pages',
      components: '.temp/components',
      assets: '.temp/assets',
      fonts: '.temp/assets/fonts',
    },
    src: {
      root: 'src',
      scripts: 'src/scripts',
      styles: 'src/sass',
      pages: 'src/pages',
      components: 'src/components',
      index: 'src/index.php',
      assets: 'src/assets',
      fonts: 'src/assets/fonts',
      vendor: 'src/vendor',
    },
    prod: {
      root: 'dist',
      scripts: 'dist/js',
      styles: 'dist/css',
      pages: 'dist/pages',
      assets: 'dist/assets',
      fonts: 'dist/assets/fonts',
      components: 'dist/components',
    }
  },
  settings: {
    dev: {
      proxy: '127.0.0.1:8010',
      port: 8080,
    },
    prod: {
      proxy: '127.0.0.1:8010',
      port: 3000,
    }
  }
};

gulp.task('scripts:build', () => {
  return gulp.src([
    `${CONFIG.paths.src.scripts}/**/*.js`,
    `!${CONFIG.paths.src.scripts}/**/*.test.js`,
  ])
    .pipe(gutil.env.env === 'production'
      ? gutil.noop()
      : sourcemaps.init())
    .pipe(babel({
      presets: ['@babel/env']
    }))
    .pipe(gutil.env.env === 'production'
      ? gutil.noop()
      : sourcemaps.write())
    .pipe(gutil.env.env === 'production'
      ? concat(CONFIG.filenames.prod.scripts)
      : concat(CONFIG.filenames.dev.scripts))
    .pipe(gutil.env.env === 'production'
      ? uglify()
      : gutil.noop())
    .pipe(gutil.env.env === 'production'
      ? rename({
        suffix: '.min'
      })
      : gutil.noop())
    .pipe(gutil.env.env === 'production'
      ? header(banner, {
        pkg: pkg
      })
      : gutil.noop())
    .pipe(gutil.env.env === 'production'
      ? gulp.dest(`${CONFIG.paths.prod.scripts}`)
      : gulp.dest(`${CONFIG.paths.dev.scripts}`))
});

gulp.task('scripts:lint', () => {
  return gulp.src(`${CONFIG.paths.src.scripts}/**/*.js`)
    .pipe(eslint())
    .pipe(eslint.format())
});

gulp.task('styles:build', () => {
  return gulp.src(`${CONFIG.paths.src.styles}/**/*.s+(a|c)ss`)
    .pipe(sass({
      outputStyle: 'nested',
      precision: 10,
      includePaths: ['.'],
      onError: console.error.bind(console, 'Sass error:')
    }))
    .pipe(autoprefixer({ browsers: AUTO_PREFIX_BROWSERS }))
    .pipe(urlAdjuster({
      prependRelative: '../',
    }))
    .pipe(gutil.env.env === 'production'
      ? concat(CONFIG.filenames.prod.styles)
      : concat(CONFIG.filenames.dev.styles))
    .pipe(gutil.env.env === 'production'
      ? header(banner, {
        pkg: pkg
      })
      : gutil.noop())
    .pipe(gutil.env.env === 'production'
      ? cleanCSS()
      : gutil.noop())
    .pipe(gutil.env.env === 'production'
      ? rename({
        suffix: '.min'
      })
      : gutil.noop())
    .pipe(gutil.env.env === 'production'
      ? gulp.dest(CONFIG.paths.prod.styles)
      : gulp.dest(CONFIG.paths.dev.styles))
    .pipe(browserSync.stream());
});

gulp.task('styles:lint', function () {
  return gulp.src(`${CONFIG.paths.src.root}/**/*.s+(a|c)ss`)
    .pipe(sassLint({
      options: {
        formatter: 'stylish',
        'merge-default-rules': false
      },
      files: { ignore: 'node_modules/!**!/!*.s+(a|c)ss' },
      rules: {
        'no-ids': 1,
        'no-mergeable-selectors': 0
      },
      configFile: '.sass-lint.yml'
    }))
    .pipe(sassLint.format())
});

gulp.task('clean', () => {
  if (gutil.env.env === 'production') {
    return del(CONFIG.paths.prod.root);
  } else if (gutil.env.env === 'development') {
    return del(CONFIG.paths.dev.root);
  } else if (gutil.env.env === 'all') {
    return del([CONFIG.paths.dev.root, CONFIG.paths.prod.root])
  } else {
    return del([CONFIG.paths.dev.root, CONFIG.paths.prod.root])
  }
});

gulp.task('move:assets', () => {
  return gulp.src([
    `${CONFIG.paths.src.assets}/**/*`,
  ], { base: CONFIG.paths.src.root })
    .pipe(gutil.env.env === 'production'
      ? imagemin()
      : gutil.noop())
    .pipe(gutil.env.env === 'production'
      ? gulp.dest(CONFIG.paths.prod.root)
      : gulp.dest(CONFIG.paths.dev.root))
});

gulp.task('move:vendor:fonts', () => {
  return gulp.src(bowerlibs('**/*.{eot,svg,ttf,woff,woff2}'))
    .pipe(gutil.env.env === 'production'
      ? gulp.dest(CONFIG.paths.prod.fonts)
      : gulp.dest(CONFIG.paths.dev.fonts)
    );
});

gulp.task('bundle:vendor', () => {
  let target = gulp.src([
    CONFIG.paths.src.index,
    `${CONFIG.paths.src.components}/**/*.{html|jade|php}`,
    `${CONFIG.paths.src.pages}/**/*.{html|jade|php}`,
  ], { base: CONFIG.paths.src.root });

  let js = gulp.src(wiredep().js);
  let css = gulp.src(wiredep().css);

  return target
    .pipe(
      inject(js.pipe(concat(gutil.env.env === 'production'
        ? CONFIG.filenames.prod.vendorJS
        : CONFIG.filenames.dev.vendorJS))
        .pipe(gutil.env.env === 'production'
          ? gulp.dest(CONFIG.paths.prod.scripts)
          : gulp.dest(CONFIG.paths.dev.scripts))
      )
    )
    .pipe(
      inject(css.pipe(concat(gutil.env.env === 'production'
        ? CONFIG.filenames.prod.vendorCSS
        : CONFIG.filenames.dev.vendorCSS))
        .pipe(urlAdjuster({
          replace: ['../fonts', '../assets/fonts'],
        }))
        .pipe(gutil.env.env === 'production'
          ? gulp.dest(CONFIG.paths.prod.styles)
          : gulp.dest(CONFIG.paths.dev.styles))
      )
    )
});

gulp.task('inject', () => {
  let target = gulp.src([
    `${CONFIG.paths.src.root}/**/*.html`,
    `${CONFIG.paths.src.root}/**/*.jade`,
    `${CONFIG.paths.src.root}/**/*.php`,
    `!${CONFIG.paths.src.vendor}/**/*`,
  ], { base: CONFIG.paths.src.root });

  let devSources = gulp.src(
    [`${CONFIG.paths.dev.scripts}/**/*.js`, `${CONFIG.paths.dev.styles}/**/*.css`, `!${CONFIG.paths.dev.scripts}/**/*.test.js`,],
    {
      read: false
    }
  );

  let devInjectionOptions = {
    ignorePath: CONFIG.paths.dev.root,
    addRootSlash: false
  };

  let prodSources = gulp.src(
    [`${CONFIG.paths.prod.scripts}/**/*.js`, `${CONFIG.paths.prod.styles}/**/*.css`, `!${CONFIG.paths.prod.scripts}/**/*.test.js`,],
    { read: false }
  );

  let prodInjectionOptions = {
    ignorePath: CONFIG.paths.prod.root,
    addRootSlash: false
  };

  return target
    .pipe(gutil.env.env === 'production'
      ? htmlmin({
        collapseWhitespace: true,
      })
      : gutil.noop())
    .pipe(
      gutil.env.env === 'production'
        ? inject(prodSources, prodInjectionOptions)
        : inject(devSources, devInjectionOptions)
    )
    .pipe(gutil.env.env === 'production'
      ? gulp.dest(CONFIG.paths.prod.root)
      : gulp.dest(CONFIG.paths.dev.root))
});

gulp.task('build',
  gulp.series(
    'move:assets', 'move:vendor:fonts', gulp.parallel('scripts:build', 'styles:build'), 'bundle:vendor', 'inject'
  ), (callback) => {
    callback();
  });

gulp.task('php', (callback) => {
  let root = CONFIG.paths.dev.root;
  let port = CONFIG.settings.dev.port;

  if (gutil.env.env === 'production') {
    root = CONFIG.paths.prod.root;
    port = CONFIG.settings.prod.port;
  }

  phpConnect.server({
    base: root,
    port: port,
    keepalive: true
  });
  callback();
});

gulp.task('browserSync', (callback) => {
  let proxy = CONFIG.settings.dev.proxy;
  let port = CONFIG.settings.dev.port;

  if (gutil.env.env === 'production') {
    proxy = CONFIG.settings.prod.proxy;
    port = CONFIG.settings.prod.port;
  }

  browserSync.init({
    proxy: proxy,
    port: port,
    open: true,
    notify: false
  });
  callback();
});

gulp.task('watch:styles', () => {
  gulp.watch(`${CONFIG.paths.src.styles}/**/*.s+(a|c)ss`, gulp.series('styles:build'))
});

gulp.task('watch:scripts', () => {
  gulp.watch(`${CONFIG.paths.src.scripts}/**/*.js`, gulp.series('scripts:build'))
    .on('change', browserSync.reload);
});

gulp.task('watch:pages', () => {
  gulp.watch([
    `${CONFIG.paths.src.root}/**/*.php`, `${CONFIG.paths.src.root}/**/*.html`, `${CONFIG.paths.src.root}/**/*.jade`,
  ], gulp.series('inject'))
    .on('change', browserSync.reload);
});

gulp.task('watch:assets', () => {
  gulp.watch(`${CONFIG.paths.src.assets}/**/*`, gulp.series('move:assets'))
    .on('change', browserSync.reload);
});

gulp.task('watch', gulp.parallel('watch:styles', 'watch:scripts', 'watch:pages', 'watch:assets'));

gulp.task('serve', gulp.series('browserSync', 'php', 'watch'));
